lib/std/special/compiler_rt/emutls.zig

```diff
@@ -0,0 +1,356 @@
+    // array (containing address of allocated variables) is thread
+    // specific and stored using pthread_setspecific().
+    //
+    size: gcc_word, // size of the object in bytes
+    alignment: gcc_word, // alignment of the object in bytes
+    object: extern union {
+        index: usize, // data[index-1] is the object address - 0 = uninit
+        address: *c_void, // object address, when in single thread env
+    },
+    default_value: ?*c_void, // null or non-zero initial value for the object
+
+    /// Helper to retrieve nad initialize global unique index per emutls variable.
+    pub inline fn get_index(self: *emutls_control) usize {
+        // Two threads could race against the same emutls_control.
+
+        // Use atomic for reading coherent value lockless.
+        const index_lockless = @atomicLoad(usize, &self.object.index, .Acquire);
+
+        if (index_lockless != 0) {
+            // index is already initialized, return it.
+            return index_lockless;
+        } else {
+            // index is uninitialized: take a lock to avoid possible race.
+            Index.lock();
+            defer Index.unlock();
+
+            const index_locked = self.object.index;
+            if (index_locked != 0) {
+                // we lost a race, but index is already initialized: nothing particular to do.
+                return index_locked;
+            }
+
+            // Store a new index atomically (for having coherent index_lockless reading).
+            @atomicStore(usize, &self.object.index, Index.new(), .Release);
+
+            return self.object.index;
+        }
+    }
+
+    /// Simple helper for testing purpose.
+    pub fn init(comptime T: type, default_value: ?*T) emutls_control {
+        return emutls_control{
+            .size = @sizeOf(T),
+            .alignment = @alignOf(T),
+            .object = .{ .index = 0 },
+            .default_value = @ptrCast(?*c_void, default_value),
+        };
+    }
+
+    /// Get the pointer on allocated storage for emutls variable.
+    pub inline fn get_pointer(self: *emutls_control, allocator: *Allocator) !*c_void {
+        // ensure ThreadStorage initialization is done
+        ThreadStorage.init_once.call();
+
+        const index = self.get_index();
+        var array = try ThreadStorage.get_array(
+            allocator,
+            index,
+        );
+
+        return try array.get_pointer(index - 1, self);
+    }
+
+    /// Testing helper for retrieving typed pointer.
+    pub fn get_typed_pointer(self: *emutls_control, comptime T: type, allocator: *Allocator) !*T {
+        assert(self.size == @sizeOf(T));
+        assert(self.alignment == @alignOf(T));
+        return @ptrCast(
+            *T,
+            @alignCast(@alignOf(*T), try self.get_pointer(allocator)),
+        );
+    }
+};
+
+// ---- tests ----
+
+const assert = std.debug.assert;
+
+test "Index" {
+    Index.lock();
+    defer Index.unlock();
+
+    assert(Index.next == 1);
```

Comment:
Use `std.testing.expect`, `assert` becomes a no-op in `ReleaseFast` and `ReleaseSmall`.